fplll/pruner/pruner.cpp

```diff
@@ -0,0 +1,307 @@
+  either version 2.1 of the License, or (at your option) any later version.
+
+  fplll is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+  GNU Lesser General Public License for more details.
+
+  You should have received a copy of the GNU Lesser General Public License
+  along with fplll. If not, see <http://www.gnu.org/licenses/>. */
+
+#include "pruner.h"
+#include "ballvol.const"
+#include "factorial.const"
+#include "fplll.h"
+
+// add components
+#include "pruner_cost.cpp"
+#include "pruner_optimize.cpp"
+#include "pruner_optimize_tc.cpp"
+#include "pruner_optimize_tp.cpp"
+#include "pruner_prob.cpp"
+#include "pruner_util.cpp"
+
+FPLLL_BEGIN_NAMESPACE
+
+// call pruner (note the float type is determined now)
+template <class FT> int run_pruner_f(ZZ_mat<mpz_t> &b, const PruningParams &param, int sel_ft)
+{
+  int gso_flags = 0;
+  if (b.get_rows() == 0 || b.get_cols() == 0)
+    return RED_SUCCESS;
+  if (sel_ft == FT_DOUBLE || sel_ft == FT_LONG_DOUBLE)
+    gso_flags |= GSO_ROW_EXPO;
+
+  // some checks
+  int start = param.prune_start;
+  int end   = param.prune_end;
+  if (start < 0 || start >= b.get_rows() - 1)
+    start = 0;
+  if (end <= start || end >= b.get_rows())
+    end                  = b.get_rows();
+  double prune_pre_nodes = param.prune_pre_nodes;
+  double prune_min_prob  = param.prune_min_prob;
+  if (prune_pre_nodes <= 1)
+    prune_pre_nodes = 1;
+  int block_size    = end - start;
+
+  PruningParams pruning;
+  vector<double> r;
+  FT root_det, max_dist;
+  long max_dist_expo;
+
+  // we check if we can convert the basis to long integers for
+  // performance
+  ZZ_mat<long> bl;
+  if (convert<long, mpz_t>(bl, b, 10))
+  {
+    ZZ_mat<long> empty_mat;
+    MatGSO<Z_NR<long>, FT> m_gso(bl, empty_mat, empty_mat, gso_flags);
+    m_gso.update_gso();
+    max_dist = m_gso.get_r_exp(start, start, max_dist_expo);
+    root_det = m_gso.get_root_det(start, end);
+    for (int i = start; i < end; ++i)
+    {
+      FT x;
+      m_gso.get_r(x, i, i);
+      r.push_back(x.get_d());
+    }
+  }
+  else
+  {
+    ZZ_mat<mpz_t> empty_mat;
+    MatGSO<Z_NR<mpz_t>, FT> m_gso(b, empty_mat, empty_mat, gso_flags);
+    m_gso.update_gso();
+    max_dist = m_gso.get_r_exp(start, start, max_dist_expo);
+    root_det = m_gso.get_root_det(start, end);
+    for (int i = start; i < end; ++i)
+    {
+      FT x;
+      m_gso.get_r(x, i, i);
+      r.push_back(x.get_d());
+    }
+  }
+
+  adjust_radius_to_gh_bound(max_dist, max_dist_expo, block_size, root_det, param.gh_factor);
+  double radius_d = max_dist.get_d() * pow(2, max_dist_expo);
+
+  cerr << "# Start Pruning" << endl;
+  cerr << "# enumeration Radius: " << radius_d << endl;
+  cerr << "# preprocessing (num. nodes): " << prune_pre_nodes << endl;
+  cerr << "# targeted min. prob: " << prune_min_prob << endl;
+  cerr << "# input GSO: " << r << endl;
```

Comment:
This looks quite noisy.

Reply:
These are outputs from the pruner. They look informative to me. Perhaps we can remove the last one. 

cerr << "# input GSO: " << r << endl;

Reply:
But doesn't that mean we cannot call the pruner without stderr being poluted with debug information? That is quite noisy, e.g. think in a multicore setting.

Reply:
The above output is only for the fplll command line usage. If someone uses ./fplll -pru in the command line, they would expect some outputs. At least the pruning parameter should be output otherwise they got no results. 

I am not sure if fplll has a multicore support. Do you mean the call from fpylll? Then the behavior should be the same as the current fpylll (there is no output except writing the result to the bkzparam).

Reply:
Ah! Sorry, my bad. All good on this one then.